msgraph/core/graph_client.py

```diff
@@ -0,0 +1,297 @@
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import json
+from typing import List, Optional
+
+import httpx
+from kiota_abstractions.authentication import AccessTokenProvider
+from kiota_http.middleware.middleware import BaseMiddleware
+
+from ._enums import APIVersion, NationalClouds
+from .graph_client_factory import GraphClientFactory
+
+
+def collect_options(func):
+    """Collect middleware options into a middleware control dict and pass it as a header"""
+
+    def wrapper(*args, **kwargs):
+
+        # These are middleware options that can be configured per request.
+        # Supports options for default middleware as well as custom middleware.
+
+        options = kwargs.pop('request_options', None)
+        if options:
+            if 'headers' in kwargs:
+                kwargs['headers'].update({'request_options': json.dumps(options)})
+            else:
+                kwargs['headers'] = {'request_options': json.dumps(options)}
+
+        return func(*args, **kwargs)
+
+    return wrapper
+
+
+class GraphClient:
```

Comment:
is inheritance from multiple parents supported in Python? If not, I'm not sure how the GraphServiceClient will inherit from this one.

Reply:
`GraphServiceClient` will be passed a `GraphRequestAdapter` instance. GraphRequestAdapter in turn accepts a `GraphClient` as an argument

Reply:
I think the naming here is leading to confusion, if this is not a client in the sense of http client or a client in the sense of a service client, what is its function? so we can find a better name for it.

Reply:
`GraphClient` is indeed a http client, and will be used to make raw string requests to the graph API by users who do not want the service library. Alse added a separate `GraphRequestAdapter` class that will be used for the full sdk/self-serve sdks.

Reply:
We may not need the separation of the two , I believe it should be possible to build an instance of `RequestInformation` using the raw string and then pass it to the `RequestAdapter` to send the request. This would mean one `GraphClient` which holds a `RequestAdapter` inside it. 
This way we can possibly avoid implementation/functional drifts due to slight variations in configuration of the client e.g Auth which is handled by the `RequestAdapter`.

Reply:
Could work. Lemme test it out. I'll revert with findings.